### DISCUSSION TITLE
RSpec Matcher for Notifications

### DISCUSSION THREAD

**Discussion by jon-sully** [excid3/noticed#318] [Ideas]
Greetings! Just wanted to put this out there for others that might be looking for it. Less for testing notifications themselves and instead for testing things that fire off notifications, validating that notifications were indeed fired off. This obviously only works for notifications that are delivered by database, such that records would actually exist to verify a notification actually got created... but it gives us a nice RSpec syntax:

```ruby
expect(my_agent).to have_received_notification BigNewsNotification
```

Neato.

Not sure whether this (or a variant of it) will become a built-in feature from this library, but it makes my tests feel a little cleaner and more specific — adding a custom matcher to your own app (instead of hoping it gets built into Noticed) is as easy as:

```ruby
# spec/support/matchers/have_received_notification.rb
RSpec::Matchers.define :have_received_notification do |expected_notification_class|
  match do |recipient|
    recipient.notifications.find_by type: expected_notification_class.name
  end

  failure_message do |recipient|
    "expected that #{recipient} would have received a #{expected_notification_class.name} notification, but did not"
  end

  failure_message_when_negated do |recipient|
    "expected that #{recipient} would not have received a #{expected_notification_class.name} notification, but did"
  end
end
```

and

```ruby
# spec/rails_helper.rb

...
require "support/matchers/have_received_notification"
...
```

This matcher will make sure that at least one notification of the given type was generated for the given recipient. The recipient must have the `has_many :notifications, as: :recipient` relationship already defined on their model.

One might quickly draft up a matcher that uses the expectation _block_ format to ensure that _only_ one notification is generated instead of this non-block-format 'at least one exists', but this was enough for me. ✌️

**Comment by excid3** [comment]
This is cool! I think we'd be happy to include this in the gem.

We're going to be starting a refactor for Noticed v2 and will have some major changes, so it will likely be best implemented after those improvements are made.

**Reply by excid3** [reply]
For Noticed 2.x, the matcher needs to join the `noticed_events`. Something like this:

```ruby
  match do |recipient|
    recipient.notifications.joins(:event).find_by(noticed_events: {type: expected_notification_class.name})
  end
```

**Comment by jon-sully** [comment]
Sounds intriguing 😁 👍🏻

**Comment by Spone** [comment]
Any update on how to test notifiers, now that v2.0 is out?

**Reply by excid3** [reply]
What are you trying to assert? If the database record was created?

**Reply by Spone** [reply]
I think I would expect a section in the README that would give an overview of the recommended testing strategy.

I think in my case, I want to:
- have a `spec/notifiers/my_notifier_spec.rb` to unit test the notifier (assert that the mailer is called with the correct params)
- implement `have_received_notification` assertions in my request tests
- test the mailer separately

**Comment by gap777** [comment]
I would love for this to have been baked right in!
